### PR TITLE
Update profiler doc (2/n)

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of docs revamp #7666

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @carmocca @kaushikb11 @ninginthecloud

### CODE REVIEW

File: docs/source/advanced/profiler.rst [hunk 1]
```diff
@@ -3,10 +3,291 @@
...
+
+Simple Profiler
+===============
+
+If you only wish to profile the standard actions, you can set ``profiler="simple"``. It uses the built-in
+:class:`~pytorch_lightning.profiler.simple.SimpleProfiler`.
+
+.. code-block:: python
+
+    # by passing a string
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:49]: Mind adding the import path of the `SimpleProfiler` ?

File: docs/source/advanced/profiler.rst [hunk 2]
```diff
@@ -3,10 +3,291 @@
...
+    trainer = Trainer(..., profiler="simple")
+
+    # or by passing an instance
+    profiler = SimpleProfiler()
+    trainer = Trainer(..., profiler=profiler)
+
+The profiler's results will be printed at the completion of a training ``trainer.fit()``. Note that there are a lot more actions that will be present in the finalized
+report. Below one is just an example containing a few:
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:57]:
```suggestion
report. Find below an example of the :class:`~pytorch_lightning.profiler.simple.SimpleProfiler` report:
```

File: docs/source/advanced/profiler.rst [hunk 3]
```diff
@@ -3,10 +3,291 @@
...
+
+.. code-block::
+
+    FIT Profiler Report
+
+    -----------------------------------------------------------------------------------------------
+    |  Action                                          |  Mean duration (s)	|  Total time (s) |
+    -----------------------------------------------------------------------------------------------
+    |  [LightningModule]BoringModel.prepare_data       |  2.112e-06      	|  2.112e-06      |
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:66]: This doesn't look entirely right. We should find the percentage and the list should be ordered based on total time.

File: docs/source/advanced/profiler.rst [hunk 4]
```diff
@@ -3,10 +3,291 @@
...
+    |  [LightningModule]BoringModel.on_fit_end         |  3.127e-06      	|  3.127e-06      |
+    |  [LightningModule]BoringModel.teardown           |  2.581e-06      	|  2.581e-06      |
+    -----------------------------------------------------------------------------------------------
+
+
+Advanced Profiler
+=================
+
+If you want more information on the functions called during each event, you can use the :class:`~pytorch_lighting.profiler.advanced.AdvancedProfiler`.
+This option uses Python's `cProfiler <https://docs.python.org/3/library/profile.html#module-cProfile>`_ to provide a report of time spent on *each* function called within your code.
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:90]:
```suggestion
This option uses Python's `cProfiler <https://docs.python.org/3/library/profile.html#module-cProfile>`_ to provide an in-depth report of time spent within *each* function called in your code.
```

File: docs/source/advanced/profiler.rst [hunk 5]
```diff
@@ -3,10 +3,291 @@
...
+
+.. code-block:: python
+
+    # by passing a string
+    trainer = Trainer(..., profiler="advanced")
+
+    # or by passing an instance
+    profiler = AdvancedProfiler()
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:98]: Add path to import the `AdvancedProfiler`.

File: docs/source/advanced/profiler.rst [hunk 6]
```diff
@@ -3,10 +3,291 @@
...
+        60000    0.182    0.000    7.020    0.000 transforms.py:93(__call__)
+        60000    1.651    0.000    6.839    0.000 functional.py:42(to_tensor)
+        60000    0.260    0.000    5.734    0.000 transforms.py:167(__call__)
+
+
+PyTorch Profiler
+================
+
+Autograd includes a profiler that lets you inspect the cost of different operators
+inside your model - both on the CPU and GPU. It uses the built-in :class:`~pytorch_lightning.profiler.pytorch.PyTorchProfiler`.
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:131]: Let's add a link to their repo.

File: docs/source/advanced/profiler.rst [hunk 7]
```diff
@@ -3,10 +3,291 @@
...
+
+    # by passing a string
+    trainer = Trainer(..., profiler="pytorch")
+
+    # or by passing an instance
+    profiler = PyTorchProfiler()
+    trainer = Trainer(..., profiler=profiler)
+
+
+This profiler works with PyTorch ``DistributedDataParallel``.
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:146]:
```suggestion
This profiler works with multi-devices training.
```

File: docs/source/advanced/profiler.rst [hunk 8]
```diff
@@ -3,10 +3,291 @@
...
+    empty                  0.27%            4.456us          0.27%            4.456us          2.228us
+    copy_                  0.15%            2.526us          0.15%            2.526us          2.526us
+    broadcast_tensors      0.15%            2.492us          0.15%            2.492us          2.492us
+    size                   0.06%            0.967us          0.06%            0.967us          0.484us
+    is_complex             0.06%            0.961us          0.06%            0.961us          0.481us
+    stride                 0.03%            0.517us          0.03%            0.517us          0.517us
+    ---------------------  ---------------  ---------------  ---------------  ---------------  ---------------
+    Self CPU time total: 1.681ms
+
+When running with ``PyTorchProfiler(emit_nvtx=True)``. You should run as following:
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:194]: Let's add we can visualize the trace in Chrome Trace (+ small giff of the steps) or tensorboard with the profiler plugin (+ small gif of the steps).

File: docs/source/advanced/profiler.rst [hunk 9]
```diff
@@ -3,10 +3,291 @@
...
+            super().teardown(stage=stage)
+
+.. code-block:: python
+
+    trainer = Trainer(..., profiler=ActionCountProfiler())
+    trainer.fit(...)
+
+
+Customize Profiling
+===================
```
Comment (tchaton) [inline docs/source/advanced/profiler.rst:270]: Neat !

File: docs/source/advanced/profiler.rst (hunk 3 above)
Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:66]: percentage and call count comes with `extended=True`(default). I tried putting it here, but it was super-ugly and nothing was clear from the output. This one is also not the original exact output. Made some alignment to render it properly in the docs.

File: docs/source/advanced/profiler.rst (hunk 6 above)
Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:131]: The profiler doc link is mentioned in 2 lines below. Is there any other reference you are referring to?

File: docs/source/advanced/profiler.rst (hunk 8 above)
Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:194]: can I add it later? I need to refine the outputs for each profiler. so if I create a gif now, I'll need to create one more again after the render is changed.

File: docs/source/advanced/profiler.rst [hunk 10]
```diff
@@ -1,81 +1,108 @@
...
+    |  [LightningModule]BoringModel.on_train_end       |  1.509e-06      	|  1.509e-06      |
+    |  [LightningModule]BoringModel.on_fit_end         |  3.127e-06      	|  3.127e-06      |
+    |  [LightningModule]BoringModel.teardown           |  2.581e-06      	|  2.581e-06      |
+    -----------------------------------------------------------------------------------------------
+
+
+Advanced Profiler
+=================
+
+If you want more information on the functions called during each event, you can use the :class:`~pytorch_lighting.profiler.advanced.AdvancedProfiler`.
```
Comment (akihironitta) [inline docs/source/advanced/profiler.rst:88]: typo :dog:
```suggestion
If you want more information on the functions called during each event, you can use the :class:`~pytorch_lightning.profiler.advanced.AdvancedProfiler`.
```

File: docs/source/advanced/profiler.rst [hunk 11]
```diff
@@ -1,81 +1,108 @@
...
-    on_batch_start          |  5.0865e-06           |  0.0095372
-    model_forward           |  0.0017818            |  3.3408
-    model_backward          |  0.0018283            |  3.4282
-    on_after_backward       |  4.2862e-06           |  0.0080366
-    optimizer_step          |  0.0011072            |  2.0759
-    on_batch_end            |  4.5202e-06           |  0.0084753
-    on_epoch_end            |  3.919e-06            |  3.919e-06
-    on_train_end            |  5.449e-06            |  5.449e-06
+    # or by passing an instance
+    from pytorch_lightning.profiler.simple import SimpleProfiler
```
Comment (akihironitta) [inline docs/source/advanced/profiler.rst:50]:
```suggestion
    from pytorch_lightning.profiler import SimpleProfiler
```

File: docs/source/advanced/profiler.rst [hunk 12]
```diff
@@ -1,81 +1,108 @@
...
+This option uses Python's `cProfiler <https://docs.python.org/3/library/profile.html#module-cProfile>`_ to provide an in-depth report of time spent within *each* function called in your code.
 
 .. code-block:: python
 
+    # by passing a string
     trainer = Trainer(..., profiler="advanced")
 
-    # or
+    # or by passing an instance
+    from pytorch_lightning.profiler.advanced import AdvancedProfiler
```
Comment (akihironitta) [inline docs/source/advanced/profiler.rst:97]:
```suggestion
    from pytorch_lightning.profiler import AdvancedProfiler
```

File: docs/source/advanced/profiler.rst [hunk 13]
```diff
@@ -97,68 +124,44 @@ of logging it to the output in your terminal. The output below shows the profili
...
+have a look at its `docs <https://pytorch.org/docs/master/profiler.html>`_.
 
 .. code-block:: python
 
+    # by passing a string
     trainer = Trainer(..., profiler="pytorch")
 
-    # or
+    # or by passing an instance
+    from pytorch_lightning.profiler.pytorch import PyTorchProfiler
```
Comment (akihironitta) [inline docs/source/advanced/profiler.rst:143]:
```suggestion
    from pytorch_lightning.profiler import PyTorchProfiler
```

**Review by akihironitta — COMMENTED** [review]
:rocket:

File: docs/source/advanced/profiler.rst [hunk 14]
```diff
@@ -191,16 +194,103 @@ The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the s
...
+            super().teardown(stage=stage)
+
+.. code-block:: python
+
+    trainer = Trainer(..., profiler=ActionCountProfiler())
+    trainer.fit(...)
+
+
+Customize Profiling
+===================
```
Comment (akihironitta) [inline docs/source/advanced/profiler.rst:273]:
```suggestion
Profile Your Own Logic
======================
```
or something like that maybe? "Customize Profiling" sounds a bit ambiguous to me.

File: docs/source/advanced/profiler.rst (hunk 10 above)
Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:88]: always happen with lightning 😅 
https://github.com/PyTorchLightning/pytorch-lightning#lightning-lite  here too 😆

File: docs/source/advanced/profiler.rst (hunk 14 above)
Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:273]: in the 2nd subsection user is not changing any logic, but adding another hook to be profiled using an existing profiler.. `Your Logic` might not be the best title. better suggestions?

**Review by akihironitta — APPROVED** [review]

File: docs/source/advanced/profiler.rst (hunk 11 above)
Comment (akihironitta) [reply docs/source/advanced/profiler.rst:50]: @rohitgr7 Would you prefer keeping this?

Comment (rohitgr7) [reply docs/source/advanced/profiler.rst]: my bad! forgot to commit. :)

File: docs/source/advanced/profiler.rst (hunk 14 above)
Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:273]: that sort of missing the first sub-seection.
took the motivation from here: https://pytorch-lightning.readthedocs.io/en/latest/common/checkpointing.html#customize-checkpointing

File: docs/source/advanced/profiler.rst [hunk 15]
```diff
@@ -1,81 +1,108 @@
...
-
 .. _profiler:
 
-Performance and Bottleneck Profiler
-===================================
+#########
+Profiling
+#########
+
+Profiling your training/testing/inference run can help you understand if there are any bottlenecks in your code. The reports can be generated with ``trainer.fit()`` ``trainer.test()``, ``trainer.validate()`` and ``trainer.predict()`` with the respective actions called.
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source/advanced/profiler.rst:7]: what does "with the respective actions called" mean?

File: docs/source/advanced/profiler.rst [hunk 16]
```diff
@@ -1,81 +1,108 @@
...
-    on_train_end            |  5.449e-06            |  5.449e-06
+    # or by passing an instance
+    from pytorch_lightning.profiler import SimpleProfiler
 
+    profiler = SimpleProfiler()
+    trainer = Trainer(..., profiler=profiler)
 
-Advanced Profiling
-------------------
+The profiler's results will be printed at the completion of a training ``trainer.fit()``. Note that there are a lot more actions that will be present in the finalized
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:55]: I would put the note below ("the above output is trunceded, more actions will be listed in the ...")

File: docs/source/advanced/profiler.rst [hunk 17]
```diff
@@ -191,16 +194,103 @@ The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the s
     ---------------------  ---------------  ---------------  ---------------  ---------------  ---------------
     Self CPU time total: 1.681ms
 
-When running with `PyTorchProfiler(emit_nvtx=True)`. You should run as following::
+When running with ``PyTorchProfiler(emit_nvtx=True)``. You should run as following:
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:197]: There is a period, let's delete it.

File: docs/source/advanced/profiler.rst [hunk 18]
```diff
@@ -191,16 +194,103 @@ The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the s
...
+----------------
+
+****************
+Custom Profiling
+****************
+
+Custom Profiler
+===============
+
+You can also configure a custom profiler and pass it to the Trainer. To configure it by subclassing :class:`~pytorch_lightning.profiler.base.BaseProfiler`.
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:223]: The sentence is broken

File: docs/source/advanced/profiler.rst [hunk 19]
```diff
@@ -191,16 +194,103 @@ The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the s
...
+The following is a simple example the profilers the first occurance and total calls to each action:
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:224]: this one too :))

File: docs/source/advanced/profiler.rst [hunk 20]
```diff
@@ -191,16 +194,103 @@ The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the s
...
+            self._action_first_occurance = {}
+            super().teardown(stage=stage)
+
+.. code-block:: python
+
+    trainer = Trainer(..., profiler=ActionCountProfiler())
+    trainer.fit(...)
+
+
+Customize Profiling
```
Comment (awaelchli) [inline docs/source/advanced/profiler.rst:272]: There are already two titles above: "custom profiling" and "custom profiler". This does not fit well here.

**Review by awaelchli — APPROVED** [review]

File: docs/source/advanced/profiler.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:7]: the report will only contain the hooks that are being called for the respective trainer methods.
I guess `for their respective actions` would be better.

File: docs/source/advanced/profiler.rst (hunk 14 above)
Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:273]: ah! you are right.. sorry I was referring to `Custom Profiling`

File: docs/source/advanced/profiler.rst [hunk 21]
```diff
@@ -191,16 +197,103 @@ The user can provide ``PyTorchProfiler(record_functions={...})`` to extend the s
...
+    trainer = Trainer(..., profiler=ActionCountProfiler())
+    trainer.fit(...)
+
+
+Profile Logic of Your Interest
+==============================
+
+You can also reference this profiler in your LightningModule to profile specific actions of interest.
+Each profiler has a method ``profile()``m which returns a context handler. Simply pass in the name of
+your action that you want to track and the profiler will record performance for code executed within this context.
```
Comment (akihironitta) [inline docs/source/advanced/profiler.rst:280]: Profilers can be used outside LightningModule, too, right?

```suggestion
You can also reference this profiler in your script to profile specific actions of interest.
Each profiler has a method ``profile()``m which returns a context handler. Simply pass in the name of
your action that you want to track and the profiler will record performance for code executed within this context.
```

Comment (rohitgr7) [reply docs/source/advanced/profiler.rst:280]: yes.. but these statements are explaining the attached example. You can refer them anywhere anyways as long as teardown is called to clear the cache.

**Review by carmocca — APPROVED** [review]
